### PR TITLE
Replace the http modifier with a more abstract CRUD modifier

### PR DESCRIPTION
**Pull request by bauer-andreas**
This is a follow up to https://github.com/Apodini/Apodini/issues/15 (@lschlesinger).
In meeting 3 we agreed on replacing the HTTP verbs with some custom CRUD verbs (and discussed to change the scope from `.environment` to `.nextComponent`).
I'm using this PR to start a discussion about a specific realization which will make its way into the main branch (by already providing a suggestion, which is currently also used in the prototype for the shared semantic model builder).

I see four points which still need to be discussed:
* How should we name it (I proposed `Operation` with this PR, but not sure if there is maybe some better name)?
* What should the scope be? The referenced issue suggest changing it to `.nextComponent` but I'm not sure if we agreed on that, thus this PR still uses `.environment`
* What should we define as the default value for the ContextKey? Setting `.READ` as the default makes sense for REST, but for gRPC it could be beneficial to tell the difference between undefined or explicitly defining `.READ` (maybe?).
* Are there any other values we should consider (which are relevant for any of the other interface exporters)?
For REST there is a difference between PUT and PATCH (overwriting the resource vs modifying individual properties of it). In that could be helpful to add another operation `.MODIFY` besides `.UPDATE`?

### CODE REVIEW

File: Sources/Apodini/Modifier/OperationModifier.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
+//
+//  OperationModifier.swift
+//  Apodini
+//
+//  Created by Paul Schmiedmayer on 6/26/20.
+//
+
+public enum Operation: RawRepresentable {
```
Comment (PSchmiedmayer) [inline Sources/Apodini/Modifier/OperationModifier.swift]: I would suggest to spell the enum cases lowercases as these are not fixed terms like `GET` for HTTP but can be expressed by verbs and `.create` looks a bit more Swift like.

In addition I would suggest to remove the conformance to `RawRepresentable` for the `Operation` enum as the raw value is currently not used and the Swift Runtime uses the enum name in print statements if there is no other information provided by a description property.

File: Sources/Apodini/Modifier/OperationModifier.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    init(_ component: ModifiedComponent, operation: Operation) {
+        self.component = component
+        self.operation = operation
+    }
+}
+
+
+extension OperationModifier: Visitable {
+    func visit(_ visitor: SynaxTreeVisitor) {
+        visitor.addContext(OperationContextKey.self, value: operation, scope: .environment) // TODO issue #15, change that to .nextComponent?
```
Comment (PSchmiedmayer) [inline Sources/Apodini/Modifier/OperationModifier.swift]: Yes, I would agree to change the operation to only affect the next component.

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
@Supereg Thanks for the improvement! I just have added a few smaller comments regarding he implementation of the `Operation` type that should make the PR even shorter, otherwise it looks great! I would also support the decision to make the `OperationModifier` only apply to the next component.

**Review by PSchmiedmayer — DISMISSED** [review]
As all tests are now passing and the changes in the `Operation` enum have been applied I approve the PR.

**Comment by bauer-andreas** [comment]
> As all tests are now passing and the changes in the `Operation` enum have been applied I approve the PR.
Thanks.

I'm still going to incorporate what we discussed in the last work meeting:
* A dedicated default `Operation`
* Some documentation detailing the behavior in the different exporters

**Review by PSchmiedmayer — DISMISSED** [review]

**Comment by PSchmiedmayer** [comment]
@Supereg Sounds good to me, then we will leave this PR open until then 👍

**Comment by bauer-andreas** [comment]
> * A dedicated default Operation
> * Some documentation detailing the behavior in the different exporters

Added both.

File: Documentation/Operation.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+As the `Operation` enum cannot be properly map to any feature of the technology itself (like HTTP methods can be
+used for HTTP), the gRPC exporter SHOULD incorporate the operation into the naming of the procedure.  
+As `.automatic` is the default value, meaning most likely the user hasn't specified any `.operation`, it should
+be ignored and no changes should be done to the procedure name.
+
+
+### GraphQL
+
+The `Operation` is not really relevant when creating the query schema. Instead the GraphQL exporter
```
Comment (PSchmiedmayer) [inline Documentation/Operation.md]: GraphQL has two different types of functions that can be exposed: Queries and Mutations. The Operation could be used to infer if a Component is exposed as a query or mutation.

**Review by PSchmiedmayer — DISMISSED** [review]
I only added a small comment regarding the GraphQL exporter, the PR otherwise looks great!

File: Documentation/Operation.md (hunk 3 above)
Comment (bauer-andreas) [reply Documentation/Operation.md]: Ah messed up here I think. What the sentence was meant to say is that `.automatic` isn't really useful to GraphQL right?
Cause GraphQL can only detect in an incoming query/mutation if that is a read or write request (or update). That information can then be used to match the appropriate `Component.handle`. Like a mutation will end up at the Component which has operation update/create.
Do I understand correctly?

So probably best would be for GraphQL to use `.read` as default for `.automatic`?

**Comment by bauer-andreas** [comment]
Reworked/fixed the instructions for the GraphQL exporter. I hope I got what you wanted in there.

As I pushed another commit you will have the re-approve the PR.

**Review by PSchmiedmayer — APPROVED** [review]

**Pull request merged**